File: android/app/src/main/java/com/app/voicenotes/ShareReceiverModule.kt
package com.app.voicenotes

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise // To return results/errors asynchronously
import com.facebook.react.modules.core.DeviceEventManagerModule

import java.io.File // For file operations
import android.util.Log

class ShareReceiverModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private val logTag = "ShareReceiverModule"

    // This is the name used to access the module from React Native
    // e.g., const { ShareReceiver } = NativeModules;
    override fun getName(): String {
        return "ShareReceiver"
    }

   @ReactMethod
    fun checkForPendingShareEvents(promise: Promise) {
        Log.d(logTag, "JS called checkForPendingShareEvents.")
        val pendingData = MainActivity.pendingShareEventData // Access static variable

        if (pendingData != null) {
            Log.d(logTag, "Pending share event data found. Attempting to send.")
            // Try to get current context - it SHOULD be ready if JS is calling this
            val currentContext = reactApplicationContext
            if (currentContext != null && currentContext.hasActiveReactInstance()) {
                 try {
                    currentContext
                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                        .emit("onShareReceived", pendingData) // Emit the queued event
                    MainActivity.pendingShareEventData = null // Clear the static variable after sending
                    Log.d(logTag, "Successfully sent pending event data.")
                    promise.resolve(true) // Indicate event was sent (or was pending)
                 } catch (e: Exception) {
                     Log.e(logTag, "Error emitting pending event", e)
                     promise.reject("E_PENDING_EMIT", "Failed to emit pending event: ${e.message}", e)
                 }
            } else {
                 Log.w(logTag, "Context not ready when trying to send pending event. Data remains queued.")
                 // Should not happen often if called from JS, but handle defensively
                 promise.reject("E_CONTEXT_NULL", "ReactContext not available when sending pending event.")
            }
        } else {
            Log.d(logTag, "No pending share event data found.")
            promise.resolve(false) // Indicate no event was pending
        }
    }

    // Method exposed to React Native to clear files stored in the app's cache directory
    @ReactMethod
    fun clearSharedFilesCache(promise: Promise) {
        try {
            val cacheDir = reactApplicationContext.cacheDir // Get app's cache directory
            if (cacheDir == null || !cacheDir.exists()) {
                Log.d(logTag, "Cache directory does not exist, nothing to clear.")
                promise.resolve(true) // Indicate success as there's nothing to do
                return
            }

            val files = cacheDir.listFiles()
            var deletedCount = 0
            var failedCount = 0

            files?.forEach { file ->
                // Optional: Add more specific logic to only delete files created by the share intent
                // (e.g., check for a prefix like "shared_") if cache is used for other things.
                try {
                    if (file.isFile && file.delete()) {
                        deletedCount++
                    } else if (file.isDirectory) {
                        // Optionally delete directories recursively if needed
                    } else if (file.isFile) { // File exists but delete failed
                        Log.w(logTag, "Failed to delete cache file: ${file.name}")
                        failedCount++
                    }
                } catch (e: SecurityException) {
                    Log.e(logTag, "SecurityException deleting cache file: ${file.name}", e)
                    failedCount++
                } catch (e: Exception) {
                    Log.e(logTag, "Exception deleting cache file: ${file.name}", e)
                    failedCount++
                }
            }

            if (failedCount == 0) {
                Log.d(logTag, "Successfully cleared $deletedCount files from cache.")
                promise.resolve(true)
            } else {
                Log.w(logTag, "Cleared $deletedCount files, but failed to delete $failedCount files from cache.")
                // Resolve partially successful, or reject based on severity
                promise.resolve(false) // Indicate partial success/failure
            }

        } catch (e: Exception) {
            Log.e(logTag, "Error accessing or clearing cache directory.", e)
            promise.reject("E_CLEAR_CACHE_ERROR", "Failed to clear cache directory: ${e.message}", e)
        }
    }

}

File: config-plugins/with-custom-share-receiver/ios/ShareReceiverModule.swift
// ShareReceiverApp/ShareReceiverModule.swift
import Foundation
import React // Ensure React is imported

// IMPORTANT: Replace with your actual App Group ID (must match Share Extension)
private let AppGroupId = "group.app.voicenotes"
// IMPORTANT: Use the SAME UserDefaults key as in the Share Extension
private let SharedItemsUserDefaultsKey = "SharedItemsKey_v2"

// Define the SAME data structure as in the Share Extension
struct SharedItemInfo: Codable {
    let type: String // "file", "text", "url"
    var fileName: String?
    var path: String?
    var mimeType: String?
    var content: String?
    let id: String

    // Add initializers if needed for internal use, but Codable handles decoding
}

@objc(ShareReceiver)
class ShareReceiver: NSObject, RCTBridgeModule {

  // Required by RCTBridgeModule
  static func moduleName() -> String! {
    return "ShareReceiver" // Name used to access this module from React Native
  }

  // Optional: Specify true if module initialization requires the main thread.
  // Usually true if accessing UIKit or main-thread-only APIs during init.
  // Accessing UserDefaults should be safe from background threads, but main thread is safer.
  static func requiresMainQueueSetup() -> Bool {
    return true
  }

  // Method exposed to React Native to check for and retrieve shared items
  @objc(checkForSharedItems:rejecter:)
  func checkForSharedItems(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) -> Void {
    NSLog("[ShareReceiverModule] checkForSharedItems method ENTERED.")

    guard let sharedDefaults = UserDefaults(suiteName: AppGroupId) else {
        NSLog("[ShareReceiverModule] ERROR: Could not access shared UserDefaults for group.")
        reject("E_NO_USER_DEFAULTS", "Could not access shared UserDefaults.", nil)
        return
    }

    // --- ** ATOMIC READ AND REMOVE LOGIC ** ---
    // 1. Attempt to read the data
    guard let data = sharedDefaults.data(forKey: SharedItemsUserDefaultsKey) else {
        // No data found, resolve empty immediately
        NSLog("[ShareReceiverModule] No data found in UserDefaults for key. Resolving empty.")
        resolve([])
        return
    }

    // 2. Data WAS found. Immediately remove it from UserDefaults.
    NSLog("[ShareReceiverModule] Found data (\(data.count) bytes) for key. REMOVING KEY NOW...")
    sharedDefaults.removeObject(forKey: SharedItemsUserDefaultsKey)
    // Optional: Verify removal if needed sharedDefaults.synchronize() though often not necessary

    // 3. Now, process the 'data' variable we read *before* removing the key.
    NSLog("[ShareReceiverModule] Key removed. Attempting to decode the retrieved data...")
    do {
        let decoder = JSONDecoder()
        let items = try decoder.decode([SharedItemInfo].self, from: data) // Decode the data read earlier
        NSLog("[ShareReceiverModule] Successfully decoded \(items.count) items.")

        // Convert to an array of dictionaries for React Native
        let result = items.map { item -> [String: Any?] in
             var dict: [String: Any?] = [ "id": item.id, "type": item.type, "mimeType": item.mimeType ]
             if item.type == "file" {
                 dict["fileName"] = item.fileName
                 dict["path"] = (item.path != nil) ? "file://" + item.path! : nil
             } else { dict["content"] = item.content }
             return dict
        }

      NSLog("[ShareReceiverModule] Resolving promise with \(result.count) items.")
        resolve(result) // Resolve with the processed data

    } catch {
      NSLog("[ShareReceiverModule] ERROR decoding shared items (after removing key)")
        reject("E_SHARE_DECODE_ERROR", "Error decoding shared item data: \(error.localizedDescription)", error)
    }
     // --- ** END OF ATOMIC READ AND REMOVE LOGIC ** ---
  }

  // Method exposed to React Native to clear the *files* stored in the App Group container
  @objc(clearSharedFilesCache:rejecter:)
   func clearSharedFilesCache(
       _ resolve: @escaping RCTPromiseResolveBlock,
       rejecter reject: @escaping RCTPromiseRejectBlock
   ) -> Void {
       let fileManager = FileManager.default
       // Get URL for the App Group container
       guard let containerURL = fileManager.containerURL(forSecurityApplicationGroupIdentifier: AppGroupId) else {
           reject("E_NO_CONTAINER", "Could not get App Group container URL for group \(AppGroupId).", nil)
           return
       }
       // Get URL for the specific sub-directory where files were saved
       let sharedFolderUrl = containerURL.appendingPathComponent("SharedFiles")

       do {
           // Check if the directory exists
           if fileManager.fileExists(atPath: sharedFolderUrl.path) {
               NSLog("[ShareReceiverModule] Found shared files directory at: \(sharedFolderUrl.path)")
               // Get list of items inside the directory
               let fileUrls = try fileManager.contentsOfDirectory(at: sharedFolderUrl, includingPropertiesForKeys: nil, options: [])
               var deletedCount = 0
               // Iterate and remove each item (file or subdirectory)
               for fileUrl in fileUrls {
                   try fileManager.removeItem(at: fileUrl)
                   deletedCount += 1
               }
               NSLog("[ShareReceiverModule] Cleared \(deletedCount) items from shared files cache directory.")
           } else {
                NSLog("[ShareReceiverModule] Shared files cache directory does not exist, nothing to clear.")
           }
           // Resolve the promise indicating success (even if directory didn't exist)
           resolve(true)
       } catch {
           // An error occurred during directory listing or file removal
           NSLog("[ShareReceiverModule] Error clearing shared files cache: \(error)")
           reject("E_CLEAR_CACHE_ERROR", "Failed to clear shared files cache: \(error.localizedDescription)", error)
       }
   }

   // Optional: If using event emitters later, you would implement methods here
   // func supportedEvents() -> [String]! { return ["onShareReceived"] } // Example
   // func startObserving() {} // Called when first listener is added
   // func stopObserving() {} // Called when last listener is removed
}
